typing/env.ml

```diff
@@ -703,20 +703,28 @@ let find_name_module ~mark name tbl =
 
 let add_persistent_structure id env =
   if not (Ident.persistent id) then invalid_arg "Env.add_persistent_structure";
-  if not (Current_unit_name.is_name_of id) then
-    let summary =
+  if not (Current_unit_name.is_name_of id) then begin
+    let material =
       match
         IdTbl.find_name wrap_module ~mark:false (Ident.name id) env.modules
       with
-      | exception Not_found | _, Mod_persistent -> env.summary
-      | _ -> Env_persistent (env.summary, id)
+      | exception Not_found | _, Mod_persistent -> false
+      | _ -> true
     in
-    { env with
-      modules = IdTbl.add id Mod_persistent env.modules;
-      summary
-    }
-  else
+    let summary =
+      if material then Env_persistent (env.summary, id)
+      else env.summary
+    in
+    let modules =
+      if material || not !Clflags.transparent_modules then
+        IdTbl.add id Mod_persistent env.modules
+      else
+        env.modules
+    in
+    { env with modules; summary }
+  end else begin
     env
+  end
```

Comment:
The code would be easier to read with the shorter case first (`if ... then env else ...`).

Reply:
Addressed